index.js

```diff
@@ -0,0 +1,563 @@
+    // charged might not get their product.
+    // This does prevent duplicate charges using the same PayPal order id.
+    throw new BedrockError(
+      'More than one Payment found for PayPal order.',
+      Errors.Duplicate, {payPalId: order.id}
+    );
+  }
+  const {total} = getTotalCost({order});
+  verifiedPurchase.totalCost = total;
+  return verifiedPurchase;
+};
+
+/**
+ * Creates a PayPal order server side with no payment
+ * information filled out.
+ *
+ * @param {object} options - Options to use.
+ * @param {Array<object>} options.purchase_units - A purchase_unit.
+ * @param {string} [options.intent = 'CAPTURE'] - The order intent.
+ *
+ * @returns {Promise<object>} The order object.
+ */
+const createGatewayPayment = async ({payment, intent = 'CAPTURE'}) => {
+  try {
+    const {id, amount: value, currency = 'USD'} = payment;
+    const amount = formatAmount({amount: {currency_code: currency, value}});
+    const purchase_units = [{
+      reference_id: id,
+      amount
+    }];
+    const {brandName, shippingPreference} = config.paypal;
+    const application_context = {
+      brand_name: brandName || 'bedrock-order',
+      shipping_preference: shippingPreference || 'NO_SHIPPING'
+    };
+    logger.debug('ORDER APPLICATION CONTEXT', {application_context});
+    const {headers, api} = await getOptions();
+    const url = `${api}/v2/checkout/orders`;
+    headers['Content-Type'] = 'application/json';
+    const options = {headers};
+    const body = {
+      intent,
+      purchase_units,
+      application_context
+    };
+    const {data} = await axios.post(url, body, options);
+    payment.paymentServiceId = data.id;
+    return {order: data, payment};
+  } catch(error) {
+    const {errorType, details} = formatAxiosError({error});
+    throw new BedrockError(
+      'Create Gateway Payment Failed.', errorType, details);
+  }
+};
+
+const processGatewayPayment = async ({payment}) => {
+  const order = await getOrderFromPayment({payment});
+  const verifiedPurchase = verifyOrder({order, payment});
+  return verifiedPurchase;
+};
+
+/**
+ * This is only used in the tests to simulate
+ * credit card failures.
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.order - A PayPal order object.
+ * @param {object} options.card - The payee's card.
+ *
+ * @returns {object} The result of the payment capture.
+ */
+const capturePaymentOrder = async ({order, body = {}}) => {
+  const {headers, api} = await getOptions();
+  const url = `${api}/v2/checkout/orders/` +
+    `${encodeURIComponent(order.id)}/capture`;
+  headers['Content-Type'] = 'application/json';
+  const options = {headers};
+  const {data} = await axios.post(url, body, options);
+  return data;
+};
+
+module.exports = {
+  id: 'PayPal',
+  type: 'paymentPlugin',
+  api: {
+    getGatewayCredentials,
+    updateGatewayPaymentAmount,
+    createGatewayPayment,
+    processGatewayPayment
+  },
+  test: {
```

Comment:
Don't think we should be combining code for testing infrastructure in the main module. We have avoided this in many of our libraries and we should continue.

Reply:
unfortunately most of the methods this depends on are private to this module so we have 2 options:

1. expose private methods in the plugin not used by anything else except one test method.
2. rewrite all the dependent private methods in a test utils file (copy and paste perhaps).